### PR TITLE
Adds checkpointing for ESM2

### PR DESCRIPTION
**Pull request by jomitchellnv**
### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
# TODO: Add code snippet
```

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

* **New Features**
  * Multi-backend distributed checkpointing (DDP, mFSDP, FSDP2) with periodic autosave, resume-from-latest, and final-model export (including safetensors).

* **Configuration**
  * New Hydra options: ckpt_dir, save_every_n_steps, resume_from_checkpoint, use_distributed_checkpoint_fsdp2; defaults lists updated to include _self_.

* **Training**
  * Per-script checkpoint directories, resumed-start-step handling to avoid re-running resumed steps, and automatic final-model export after training.

* **Tests**
  * Extensive end-to-end checkpoint/resume and final-model tests across backends with multi-GPU gating; improved teardown and CUDA cleanup.

### CODE REVIEW

File: bionemo-recipes/recipes/esm2_native_te/train_ddp.py [hunk 1]
```diff
@@ -56,12 +65,9 @@ def main(args: DictConfig) -> float | None:
     if dist_config.is_main_process():
         wandb.init(**args.wandb_init_args, config=OmegaConf.to_container(args, resolve=True, throw_on_missing=True))
 
-    # Create an empty ESM-2 model with a masked language model head.
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_ddp.py:61]: Any reason to remove these?

**Review by pstjohn — APPROVED** [review]

File: bionemo-recipes/recipes/esm2_native_te/train_ddp.py (hunk 1 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_ddp.py:61]: accidental during rebase

File: bionemo-recipes/recipes/esm2_native_te/checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    # Only rank 0 saves the checkpoint
+    if not dist_config.is_main_process():
+        return
+
+    checkpoint_path = os.path.join(ckpt_dir, f"step_{step}.pt")
+    os.makedirs(ckpt_dir, exist_ok=True)
+
+    torch.save({"model": model_state_dict, "optimizer": optimizer_state_dict, "step": step}, checkpoint_path)
+    logger.info(f"Saved FSDP2 checkpoint to {checkpoint_path}")
+
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py]: i mean yea ill try the patch if you have it.

[the remark this answers is not in the document]
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/checkpoint.py]: oof ok yea this was a big change -- one second. But also nice find.

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py [hunk 3]
```diff
@@ -116,9 +128,31 @@ def main(args: DictConfig) -> float | None:  # noqa: C901
...
+        )
+        # Increment start_step to avoid re-running the checkpointed step
+        start_step = min(start_step + 1, args.num_train_steps)
+        # Align LR scheduler to start at the correct step
+        try:
+            scheduler.last_epoch = start_step - 1
+            scheduler.step()
+        except Exception:
+            for _ in range(start_step):
+                scheduler.step()
```
[the remark this answers is not in the document]
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:150]: fixed
